### PR TITLE
Sets enhancements, fixes #2467

### LINKED ISSUE #2467: Add excl for OrderedSet
OrderedSet is lacking `excl` procedure at the moment

### PR DESCRIPTION
**Pull request by GULPF**
- Added excl(key) to OrderedSet
- Added pairs() iterator to OrderedSet

I noticed that `proc incl*[A](s: var HashSet[A], other: OrderedSet[A])` is a bit strange, since it takes one of each set type instead of two of the same like `proc incl*[A](s: var HashSet[A], other: HashSet[A])`. The doc-comment implies that the first is incorrect.

### CODE REVIEW

File: lib/pure/collections/sets.nim [hunk 1]
```diff
@@ -293,8 +293,6 @@ proc excl*[A](s: var HashSet[A], key: A) =
   var i = rawGet(s, key, hc)
   var msk = high(s.data)
   if i >= 0:
-    s.data[i].hcode = 0
```
Comment (Araq) [inline lib/pure/collections/sets.nim]: Why was that removed?

File: lib/pure/collections/sets.nim [hunk 2]
```diff
@@ -293,8 +293,6 @@ proc excl*[A](s: var HashSet[A], key: A) =
...
-    s.data[i].key = default(type(s.data[i].key))
```
Comment (Araq) [inline lib/pure/collections/sets.nim]: Why was that removed?

File: lib/pure/collections/sets.nim (hunk 1 above)
Comment (GULPF) [reply lib/pure/collections/sets.nim]: Those assignments are redundant since the same assignments will be done in the first loop iteration anyways. I didn't include them for the new OrderedSet.excl, so I removed them from HashSet.excl for consistency.

**Comment by GULPF** [comment]
After looking at this PR again, I feel like Nim would gain a lot by having an `Iterable[T]` concept in the standard library. Procs like `proc excl*[A](s: var HashSet[A], other: HashSet[A])` and `proc incl*[A](s: var HashSet[A], other: HashSet[A])` would be a lot more useful if the second argument was such a concept. And then it wouldn't break when somebody wants to `incl` an OrderedSet into an ordinary HashSet.

**Comment by Yardanico** [comment]
@GULPF well I can't get this: https://github.com/nim-lang/Nim/issues/6277 to work.
and without arrays iterable concept isn't very useful :)

**Comment by GULPF** [comment]
Rebased and added `missingOrExcl`
